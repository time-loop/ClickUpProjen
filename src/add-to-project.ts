import { typescript, YamlFile } from 'projen';

export module addToProjectWorkflow {
  export const RENOVATE_GITHUB_USERNAME = 'cu-infra-svc-git';

  export const DEFAULT_RENOVATE_PR_LABEL = 'renovate';
  export const GITHUB_PROJECT_NUMBER = '3';

  const defaultWorkflow = {
    name: 'add-to-project',
    on: {
      pull_request: {
        types: ['opened', 'labeled'],
      },
    },
    // Prevent renovate running multiple times in parallel due to several subsequent trigger events occurring
    concurrency: {
      group: "${{ github.workflow }}-${{ github.event_name == 'workflow_dispatch' && github.event_name || '' }}",
    },
    jobs: {
      'add-to-project': {
        'runs-on': 'ubuntu-latest',
        permissions: {
          contents: 'read',
        },
        steps: [
          {
            name: 'Checkout',
            uses: 'actions/checkout@v3',
          },
          {
            name: 'Add issue to project',
            uses: 'actions/add-to-project@v0.4.1',
            with: {
              // github project url
              'project-url': `https://github.com/orgs/time-loop/projects/${GITHUB_PROJECT_NUMBER}`,
              'github-token': 'x-access-token:${{ secrets.PROJEN_GITHUB_TOKEN }}',
              labeled: DEFAULT_RENOVATE_PR_LABEL,
            },
          },
        ],
      },
    },
  };

  export function addAddToProjectWorkflowYml(project: typescript.TypeScriptProject, override?: any): void {
    new YamlFile(project, '.github/workflows/add-to-project.yml', { obj: { ...defaultWorkflow, ...override } });
  }
}
