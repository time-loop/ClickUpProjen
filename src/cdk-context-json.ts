import { awscdk, JsonPatch, SampleFile } from 'projen';

export module cdkContextJson {
  export interface InjectionOptions {
    /**
     * The account number where the role the GH actions will assume lives.
     * This is usually the same account running cdk-pipelines,
     * but can be any account which has been configured with
     *
     * ```ts
     * cdk bootstrap --trust-for-lookup $this_account_number ...
     * ```
     */
    readonly lookupAccountId: string;
    /**
     * Which region should GH Auth into?
     * @default 'us-west-2'
     */
    readonly awsRegion?: string;
    /**
     * How long should the requested role be valid.
     * @default 900 15min seems pretty generous.
     */
    readonly roleDurationSeconds?: number;
  }

  /**
   * Options to support cdk.context.json management
   */
  export interface Options {
    /**
     * @default - do not create stack SampleCode for OIDC role
     */
    readonly createOidcRoleStack?: boolean;
    /**
     * You must configure this if you want to self-mutation cdk.context.json.
     *
     * @default - do not modify the build workflow to authorize against AWS.
     */
    readonly injectionOptions?: InjectionOptions;
  }

  export function injectAwsAuthIntoBuild(project: awscdk.AwsCdkTypeScriptApp, options: InjectionOptions): void {
    const build = project.tryFindObjectFile('.github/workflows/build.yml');
    build?.addOverride('jobs.build.permissions', { 'id-token': 'write' });
    build?.patch(
      JsonPatch.add('/jobs/build/steps/0', {
        name: 'Configure AWS Credentials',
        uses: 'aws-actions/configure-aws-credentials@v2',
        with: {
          'aws-region': options.awsRegion ?? 'us-west-2',
          'role-to-assume': `arn:aws:iam::${options.lookupAccountId}:role/${project.name}-github-actions-role`,
          'role-duration-seconds': options.roleDurationSeconds ?? 900,
        },
      }),
    );
  }

  export function addOidcRoleStack(project: awscdk.AwsCdkTypeScriptApp): void {
    new SampleFile(project, `${project.srcdir}/github-actions-oidc-cdk-context-lookup-role.ts`, {
      contents: `import { core } from '@time-loop/cdk-library';
import { aws_iam, Stage } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { Namer } from 'multi-convention-namer';

export class GitHubActionsOIDCCdkContextLookupRole extends core.Stack {
  static asStage(scope: Construct, stageName: string, stageProps: core.StageProps): Stage {
    return new (class extends Stage {
      constructor() {
        super(scope, stageName, stageProps);
        new GitHubActionsOIDCCdkContextLookupRole(this, stageProps);
      }
    })();
  }
  constructor(scope: Construct, props: core.StackProps) {
    const projectName = '${project.name}';
    // NOTE: you should never deploy this anywhere but the cdkPipelines account
    // And the GitHubActionsOIDCPermissions class for cdk-diff should never be deployed
    // in the cdkPipelines account. Name collision is intentional here.
    let id = new Namer([...projectName.split('-'), 'github', 'actions']);
    super(scope, id, props);

    const githubActionsRoleName = id.addSuffix(['role']).kebab;
    const githubActionsRole = new aws_iam.Role(this, githubActionsRoleName, {
      roleName: githubActionsRoleName,
      assumedBy: new aws_iam.FederatedPrincipal(
        \`arn:aws:iam::\${this.account}:oidc-provider/token.actions.githubusercontent.com\`,
        {
          StringEquals: {
            'token.actions.githubusercontent.com:aud': 'sts.amazonaws.com',
          },
          StringLike: {
            'token.actions.githubusercontent.com:sub': \`repo:time-loop/\${projectName}:*\`,
          },
        },
        'sts:AssumeRoleWithWebIdentity',
      ),
    });

    const githubActionsPolicyName = id.addSuffix(['policy']).kebab;
    const githubActionsPolicy = new aws_iam.Policy(this, githubActionsPolicyName, {
      policyName: githubActionsPolicyName,
      statements: [
        new aws_iam.PolicyStatement({
          actions: [ 'sts:AssumeRole' ],
          resources: [ 'arn:aws:iam::*:role/cdk-*lookup-role*' ],
        }),
      ],
    });

    // Attach IAM policy to IAM role
    githubActionsPolicy.attachToRole(githubActionsRole);
  }
}
`,
    });
  }
}
