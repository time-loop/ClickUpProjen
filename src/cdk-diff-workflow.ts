import { typescript, YamlFile } from 'projen';
import { NodePackage } from 'projen/lib/javascript';

export module cdkDiffWorkflow {
  function createCdkDiffWorkflow(options: CDKDiffOptionsConfig) {
    const defaultWorkflow = {
      name: 'cdk-diff',
      on: {
        pull_request: {},
        workflow_dispatch: {},
      },
      permissions: {
        'id-token': 'write',
        contents: 'write',
        'pull-requests': 'write',
      },
      jobs: {
        diff: {
          'runs-on': 'ubuntu-latest',
          env: {
            CI: 'true',
            GITHUB_PR_NUMBER: '${{github.event.pull_request.number}}',
            REPO_NAME: '${{ github.event.repository.name }}',
          },
          steps: [
            {
              name: 'Checkout',
              uses: 'actions/checkout@v3',
              with: {
                ref: '${{ github.event.pull_request.head.ref }}',
                repository: '${{ github.event.pull_request.head.repo.full_name }}',
              },
            },
            {
              name: 'GitHub Packages authorization',
              run: 'cat > .npmrc <<EOF\n//npm.pkg.github.com/:_authToken=${{ secrets.ALL_PACKAGE_READ_TOKEN }}\n@time-loop:registry=https://npm.pkg.github.com/\nEOF',
            },
            {
              name: 'Setup Node.js',
              uses: 'actions/setup-node@v3',
              with: {
                'node-version': '14.x',
              },
            },
            {
              name: 'Install dependencies',
              run: 'yarn install --check-files',
            },
            {
              name: 'ls stacks in each env',
              run: './node_modules/.bin/cdk ls -l -j > cdk-ls.json\nQA_STACKS=$(jq -r \'[.[].id | select(contains("Qa")) ] | join(" ")\' cdk-ls.json)\nSTG_STACKS=$(jq -r \'[.[].id | select(contains("Staging")) ] | join(" ")\' cdk-ls.json)\nPROD_STACKS=$(jq -r \'[.[].id | select(contains("Prod")) ] | join(" ")\' cdk-ls.json)\necho "QA_STACKS=$QA_STACKS" >> $GITHUB_ENV\necho "STG_STACKS=$STG_STACKS" >> $GITHUB_ENV\necho "PROD_STACKS=$PROD_STACKS" >> $GITHUB_ENV\n',
            },
            {
              name: 'configure qa aws credentials',
              uses: 'aws-actions/configure-aws-credentials@v1',
              with: {
                'role-to-assume': options.oidcQaRoleArn,
                'role-duration-seconds': 900,
                'aws-region': 'us-west-2',
              },
            },
            {
              name: 'diff qa',
              run: "set -o pipefail\n./node_modules/.bin/cdk diff $QA_STACKS --progress=events --staging false -e --ignore-errors --version-reporting false &> >(tee cdk-qa.log)\n./node_modules/.bin/ts-node ./node_modules/@time-loop/cdk-log-parser/src/cdkLogParser.ts cdk-qa.log && echo 'QA_HAS_NO_DIFF=true' >> $GITHUB_ENV\n",
            },
            {
              name: 'configure staging aws credentials',
              uses: 'aws-actions/configure-aws-credentials@v1',
              with: {
                'role-to-assume': options.oidcStagingRoleArn,
                'role-duration-seconds': 900,
                'aws-region': 'us-west-2',
              },
            },
            {
              name: 'diff staging',
              run: "set -o pipefail\n./node_modules/.bin/cdk diff $STG_STACKS --progress=events --staging false -e --ignore-errors --version-reporting false &> >(tee cdk-staging.log)\n./node_modules/.bin/ts-node ./node_modules/@time-loop/cdk-log-parser/src/cdkLogParser.ts cdk-staging.log && echo 'STAGING_HAS_NO_DIFF=true' >> $GITHUB_ENV\n",
            },
            {
              name: 'configure prod aws credentials',
              uses: 'aws-actions/configure-aws-credentials@v1',
              with: {
                'role-to-assume': options.oidcProdRoleArn,
                'role-duration-seconds': 900,
                'aws-region': 'us-west-2',
              },
            },
            {
              name: 'diff prod',
              run: "set -o pipefail\n./node_modules/.bin/cdk diff $PROD_STACKS --progress=events --staging false -e --ignore-errors --version-reporting false &> >(tee cdk-prod.log)\n./node_modules/.bin/ts-node ./node_modules/@time-loop/cdk-log-parser/src/cdkLogParser.ts cdk-prod.log && echo 'PROD_HAS_NO_DIFF=true' >> $GITHUB_ENV\n",
            },
            {
              name: 'cdk-notify',
              run: "curl -L \"https://github.com/karlderkaefer/cdk-notifier/releases/download/v2.3.2/cdk-notifier_$(uname)_amd64.gz\" -o cdk-notifier.gz\ngunzip cdk-notifier.gz && chmod +x cdk-notifier && rm -rf cdk-notifier.gz\n./cdk-notifier --log-file ./cdk-qa.log --repo $REPO_NAME --owner $GITHUB_REPOSITORY_OWNER --token ${{ secrets.GITHUB_TOKEN }} --pull-request-id ${{ github.event.pull_request.number }} -t 'Qa Stacks'\n./cdk-notifier --log-file ./cdk-staging.log --repo $REPO_NAME --owner $GITHUB_REPOSITORY_OWNER --token ${{ secrets.GITHUB_TOKEN }} --pull-request-id ${{ github.event.pull_request.number }} -t 'Staging Stacks'\n./cdk-notifier --log-file ./cdk-prod.log --repo $REPO_NAME --owner $GITHUB_REPOSITORY_OWNER --token ${{ secrets.GITHUB_TOKEN }} --pull-request-id ${{ github.event.pull_request.number }} -t 'Production Stacks'\n",
            },
            {
              name: 'Apply qa has no diff label based on diff status',
              if: "env.QA_HAS_NO_DIFF == 'true'",
              uses: 'actions/github-script@v6',
              with: {
                'github-token': '${{ secrets.GITHUB_TOKEN }}',
                script:
                  'github.rest.issues.addLabels({\n  issue_number: context.issue.number,\n  owner: context.repo.owner,\n  repo: context.repo.repo,\n  labels: ["no-qa-changes"]\n})\n',
              },
            },
            {
              name: 'Apply staging has no diff label based on diff status',
              if: "env.STAGING_HAS_NO_DIFF == 'true'",
              uses: 'actions/github-script@v6',
              with: {
                'github-token': '${{ secrets.GITHUB_TOKEN }}',
                script:
                  'github.rest.issues.addLabels({\n  issue_number: context.issue.number,\n  owner: context.repo.owner,\n  repo: context.repo.repo,\n  labels: ["no-staging-changes"]\n})\n',
              },
            },
            {
              name: 'Apply prod has no diff label based on diff status',
              if: "env.PROD_HAS_NO_DIFF == 'true'",
              uses: 'actions/github-script@v6',
              with: {
                'github-token': '${{ secrets.GITHUB_TOKEN }}',
                script:
                  'github.rest.issues.addLabels({\n  issue_number: context.issue.number,\n  owner: context.repo.owner,\n  repo: context.repo.repo,\n  labels: ["no-prod-changes"]\n})\n',
              },
            },
            {
              name: 'Should auto merge',
              id: 'should-auto-merge',
              if: "env.QA_HAS_NO_DIFF == 'true' && env.STAGING_HAS_NO_DIFF == 'true' && env.PROD_HAS_NO_DIFF == 'true' && github.event.pull_request.user.login == 'cu-infra-svc-git'",
              run: 'echo "auto-merge=true" >> $GITHUB_OUTPUT',
            },
            {
              name: 'Enable automerge',
              uses: 'daneden/enable-automerge-action@v1.0.2',
              if: 'steps.should-auto-merge.outputs.auto-merge',
              with: {
                'github-token': '${{ secrets.PROJEN_GITHUB_TOKEN }}',
                'merge-method': 'SQUASH',
                'allowed-author': 'cu-infra-svc-git',
              },
            },
            {
              name: 'Auto approve',
              uses: 'hmarr/auto-approve-action@v2.2.1',
              if: 'steps.should-auto-merge.outputs.auto-merge',
              with: {
                'github-token': '${{ secrets.GITHUB_TOKEN }}',
              },
            },
          ],
        },
      },
    };

    return defaultWorkflow;
  }

  export interface CDKDiffOptionsConfig {
    /**
     * Name of the QA OIDC role name which contains neccesasry IAM policies to run the CDK diff
     *
     * Example arn: `arn:aws:iam::123123123123:role/squad-github-actions-permissions-squad-cdk-github-actions-role`
     */
    readonly oidcQaRoleArn: string;

    /**
     * Name of the Staging OIDC role name which contains neccesasry IAM policies to run the CDK diff
     *
     * Example arn: `arn:aws:iam::123123123123:role/squad-github-actions-permissions-squad-cdk-github-actions-role`
     */
    readonly oidcStagingRoleArn: string;

    /**
     * Name of the Prod OIDC role name which contains neccesasry IAM policies to run the CDK diff
     *
     * Example arn: `arn:aws:iam::123123123123:role/squad-github-actions-permissions-squad-cdk-github-actions-role`
     */
    readonly oidcProdRoleArn: string;
  }

  export function getCDKDiffOptions(options?: CDKDiffOptionsConfig) {
    return options;
  }

  export function addCdkDiffWorkflowYml(
    project: typescript.TypeScriptProject,
    options: CDKDiffOptionsConfig,
    override?: any,
  ): void {
    new YamlFile(project, '.github/workflows/cdk-diff.yml', {
      obj: { ...createCdkDiffWorkflow(options), ...override },
    });
  }

  export function AddCdkLogParserDependency(pkg: NodePackage) {
    pkg.addDevDeps('@time-loop/cdk-log-parser@0.0.0');
  }
}
