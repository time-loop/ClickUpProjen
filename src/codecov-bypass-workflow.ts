import { YamlFile } from 'projen';
import { clickupCdk } from './clickup-cdk';

const CODECOV_GITHUB_APP_ID = 254;

export module codecovBypassWorkflow {
  function createCodecovBypassWorkflow(options?: CodecovBypassOptionsConfig) {
    const defaultWorkflow = {
      name: options?.workflowName || 'Code coverage increased',
      on: {
        check_suite: {
          types: ['completed'],
        },
        pull_request: {
          types: ['labeled', 'unlabeled'],
        },
      },
      permissions: {},
      jobs: {
        'code-coverage-increased': {
          'runs-on': 'ubuntu-latest',
          if: `github.event_name != \'check_suite\' || github.event.check_suite.app.id == ${CODECOV_GITHUB_APP_ID}`,
          steps: [
            {
              name: 'Check codecov results',
              uses: 'time-loop/github-actions/dist/wrap-check-suite@wrap-check-suite+0.3.2',
              env: {
                FORCE_COLOR: 3,
              },
              with: {
                'github-app-id': options?.githubAppId || '${{ vars.CODECOV_GITHUB_APP_ID }}',
                'github-private-key': options?.githubAppPrivateKey || '${{ secrets.CODECOV_GITHUB_APP_PRIVATE_KEY }}',
                'skip-label': options?.skipLabel || 'code coverage not required',
                'check-name': options?.checkName || 'Code coverage increased',
                'check-suite-app-id': CODECOV_GITHUB_APP_ID,
                'check-suite-check-names': options?.checkSuiteCheckNames?.join(',') || 'codecov/patch,codecov/project',
                'details-url':
                  options?.detailsUrl || 'https://app.codecov.io/gh/${{ github.repository }}/pull/<% prNumber %>',
              },
            },
          ],
        },
      },
    };

    return defaultWorkflow;
  }

  export interface CodecovBypassOptionsConfig {
    /**
     * GitHub App ID
     * @default '${{ vars.CODECOV_GITHUB_APP_ID }}'
     */
    readonly githubAppId?: string;

    /**
     * GitHub App Private Key
     * @default '${{ secrets.CODECOV_GITHUB_APP_PRIVATE_KEY }}'
     */
    readonly githubAppPrivateKey?: string;

    /**
     * Skip creating (using) the workflow
     * @default false
     */
    readonly disabled?: boolean;

    /**
     * Workflow Name
     * @default 'Code coverage increased'
     */
    readonly workflowName?: string;

    /**
     * Skip Label
     * @default 'code coverage not required'
     */
    readonly skipLabel?: string;

    /**
     * Check Name
     * @default 'Code coverage increased'
     */
    readonly checkName?: string;

    /**
     * Check Suite Check Names
     * @default ['codecov/patch', 'codecov/project']
     */
    readonly checkSuiteCheckNames?: string[];

    /**
     * Details URL
     * @default 'https://app.codecov.io/gh/${{ github.repository }}/pull/<% prNumber %>'
     */
    readonly detailsUrl?: string;
  }

  export function getCodecovBypassOptions(options?: CodecovBypassOptionsConfig) {
    return options;
  }

  export function addCodecovBypassWorkflowYml(
    project: clickupCdk.ClickUpCdkTypeScriptApp,
    options?: CodecovBypassOptionsConfig,
    override?: any,
  ): void {
    if (options?.disabled) {
      return;
    }
    new YamlFile(project, '.github/workflows/codecov-bypass.yml', {
      obj: { ...createCodecovBypassWorkflow(options), ...override },
    });
  }
}
