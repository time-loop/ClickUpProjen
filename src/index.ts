export * from './clickup-cdk';
export * from './clickup-ts';
export * from './datadog';
export * from './slack-alert';
export * from './renovate-workflow';
export * from './cdk-diff-workflow';
export * from './add-to-project';
