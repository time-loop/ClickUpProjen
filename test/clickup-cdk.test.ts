import { Testing } from 'projen';
import { clickupCdk } from '../src';

describe('ClickUpCdkTypeScriptApp', () => {
  describe('defaults', () => {
    const p = new clickupCdk.ClickUpCdkTypeScriptApp({
      name: 'test',
      cdkVersion: '2.1.0',
      defaultReleaseBranch: 'main',
    });
    const synth = Testing.synth(p);
    ['README.md', 'package.json', 'src/main.ts', 'src/widget.ts', 'test/widget.test.ts'].forEach((file) => {
      test(`snapshot ${file}`, () => {
        expect(synth[file]).toMatchSnapshot();
      });
    });

    test('prettier is enabled', () => {
      expect(p.prettier).toBeTruthy();
    });

    test('jest is enabled', () => {
      expect(p.jest).toBeTruthy();
    });

    // TODO: soooo many more tests need to be written here.
  });
});
