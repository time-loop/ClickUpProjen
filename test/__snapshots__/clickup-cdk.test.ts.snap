// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClickUpCdkConstructLibrary defaults package.json 1`] = `
Object {
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
  "author": Object {
    "email": "devops@clickup.com",
    "name": "ClickUp",
    "organization": true,
  },
  "devDependencies": Object {
    "@types/jest": "^27",
    "@types/node": "^14",
    "@typescript-eslint/eslint-plugin": "^5",
    "@typescript-eslint/parser": "^5",
    "aws-cdk-lib": "2.1.0",
    "constructs": "10.0.5",
    "esbuild": "*",
    "eslint": "^8",
    "eslint-config-prettier": "*",
    "eslint-import-resolver-node": "*",
    "eslint-import-resolver-typescript": "*",
    "eslint-plugin-import": "*",
    "eslint-plugin-prettier": "*",
    "jest": "^27",
    "jest-junit": "^13",
    "jsii": "*",
    "jsii-diff": "*",
    "jsii-docgen": "*",
    "jsii-pacmak": "*",
    "jsii-release": "*",
    "json-schema": "*",
    "npm-check-updates": "^15",
    "prettier": "*",
    "projen": "*",
    "standard-version": "^9",
    "ts-jest": "^27",
    "ts-node": "^10",
    "typescript": "*",
  },
  "engines": Object {
    "node": ">= 14.18.0",
  },
  "jest": Object {
    "clearMocks": true,
    "collectCoverage": true,
    "collectCoverageFrom": Array [
      "src/**/*.ts",
    ],
    "coverageDirectory": "coverage",
    "coveragePathIgnorePatterns": Array [
      "/node_modules/",
    ],
    "coverageReporters": Array [
      "json",
      "lcov",
      "clover",
      "cobertura",
      "text",
    ],
    "globals": Object {
      "ts-jest": Object {
        "tsconfig": "tsconfig.dev.json",
      },
    },
    "preset": "ts-jest",
    "reporters": Array [
      "default",
      Array [
        "jest-junit",
        Object {
          "outputDirectory": "test-reports",
        },
      ],
    ],
    "testMatch": Array [
      "<rootDir>/src/**/__tests__/**/*.ts?(x)",
      "<rootDir>/(test|src)/**/*(*.)@(spec|test).ts?(x)",
    ],
    "testPathIgnorePatterns": Array [
      "/node_modules/",
    ],
    "watchPathIgnorePatterns": Array [
      "/node_modules/",
    ],
  },
  "jsii": Object {
    "outdir": "dist",
    "targets": Object {},
    "tsc": Object {
      "outDir": "lib",
      "rootDir": "src",
    },
  },
  "keywords": Array [
    "cdk",
  ],
  "license": "UNLICENSED",
  "main": "lib/index.js",
  "name": "@time-loop/test",
  "peerDependencies": Object {
    "aws-cdk-lib": "^2.1.0",
    "constructs": "^10.0.5",
  },
  "publishConfig": Object {
    "registry": "https://npm.pkg.github.com/",
  },
  "repository": Object {
    "type": "git",
    "url": "https://github.com/time-loop/test.git",
  },
  "resolutions": Object {
    "@types/prettier": "2.6.0",
  },
  "scripts": Object {
    "build": "npx projen build",
    "bump": "npx projen bump",
    "clobber": "npx projen clobber",
    "compat": "npx projen compat",
    "compile": "npx projen compile",
    "default": "npx projen default",
    "docgen": "npx projen docgen",
    "eject": "npx projen eject",
    "eslint": "npx projen eslint",
    "package": "npx projen package",
    "package-all": "npx projen package-all",
    "package:js": "npx projen package:js",
    "post-compile": "npx projen post-compile",
    "post-upgrade": "npx projen post-upgrade",
    "pre-compile": "npx projen pre-compile",
    "projen": "npx projen",
    "release": "npx projen release",
    "test": "npx projen test",
    "test:update": "npx projen test:update",
    "test:watch": "npx projen test:watch",
    "unbump": "npx projen unbump",
    "upgrade": "npx projen upgrade",
    "watch": "npx projen watch",
  },
  "stability": "stable",
  "types": "lib/index.d.ts",
  "version": "0.0.0",
}
`;

exports[`ClickUpCdkTypeScriptApp defaults README.md 1`] = `"# replace this"`;

exports[`ClickUpCdkTypeScriptApp defaults package.json 1`] = `
Object {
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
  "author": Object {
    "name": "ClickUp",
    "organization": true,
  },
  "dependencies": Object {
    "@time-loop/cdk-library": "*",
    "@time-loop/cdk-named-environments": "*",
    "@time-loop/clickup-projen": "*",
    "aws-cdk-lib": "^2.1.0",
    "cdk-constants": "*",
    "constructs": "^10.0.5",
    "multi-convention-namer": "*",
  },
  "devDependencies": Object {
    "@types/jest": "^27",
    "@types/node": "^14",
    "@typescript-eslint/eslint-plugin": "^5",
    "@typescript-eslint/parser": "^5",
    "aws-cdk": "^2.1.0",
    "esbuild": "*",
    "eslint": "^8",
    "eslint-config-prettier": "*",
    "eslint-import-resolver-node": "*",
    "eslint-import-resolver-typescript": "*",
    "eslint-plugin-import": "*",
    "eslint-plugin-prettier": "*",
    "jest": "^27",
    "jest-junit": "^13",
    "jsii-release": "*",
    "json-schema": "*",
    "npm-check-updates": "^15",
    "prettier": "*",
    "projen": "*",
    "standard-version": "^9",
    "ts-jest": "^27",
    "ts-node": "^10",
    "typescript": "*",
  },
  "engines": Object {
    "node": ">= 14.18.0",
  },
  "jest": Object {
    "clearMocks": true,
    "collectCoverage": true,
    "collectCoverageFrom": Array [
      "src/**/*.ts",
    ],
    "coverageDirectory": "coverage",
    "coveragePathIgnorePatterns": Array [
      "/node_modules/",
      "/src/main.ts",
    ],
    "coverageReporters": Array [
      "json",
      "lcov",
      "clover",
      "cobertura",
      "text",
    ],
    "globals": Object {
      "ts-jest": Object {
        "tsconfig": "tsconfig.dev.json",
      },
    },
    "preset": "ts-jest",
    "reporters": Array [
      "default",
      Array [
        "jest-junit",
        Object {
          "outputDirectory": "test-reports",
        },
      ],
    ],
    "testMatch": Array [
      "<rootDir>/src/**/__tests__/**/*.ts?(x)",
      "<rootDir>/(test|src)/**/*(*.)@(spec|test).ts?(x)",
    ],
    "testPathIgnorePatterns": Array [
      "/node_modules/",
    ],
    "watchPathIgnorePatterns": Array [
      "/node_modules/",
    ],
  },
  "license": "UNLICENSED",
  "name": "test",
  "publishConfig": Object {
    "registry": "https://npm.pkg.github.com/",
  },
  "scripts": Object {
    "build": "npx projen build",
    "bump": "npx projen bump",
    "bundle": "npx projen bundle",
    "clobber": "npx projen clobber",
    "compile": "npx projen compile",
    "default": "npx projen default",
    "deploy": "npx projen deploy",
    "destroy": "npx projen destroy",
    "diff": "npx projen diff",
    "eject": "npx projen eject",
    "eslint": "npx projen eslint",
    "package": "npx projen package",
    "post-compile": "npx projen post-compile",
    "post-upgrade": "npx projen post-upgrade",
    "pre-compile": "npx projen pre-compile",
    "projen": "npx projen",
    "release": "npx projen release",
    "synth": "npx projen synth",
    "synth:silent": "npx projen synth:silent",
    "test": "npx projen test",
    "test:update": "npx projen test:update",
    "test:watch": "npx projen test:watch",
    "unbump": "npx projen unbump",
    "upgrade": "npx projen upgrade",
    "watch": "npx projen watch",
  },
  "version": "0.0.0",
}
`;

exports[`ClickUpCdkTypeScriptApp defaults src/main.ts 1`] = `
"import { core } from '@time-loop/cdk-library';
import { App } from 'aws-cdk-lib';
import { Namer } from 'multi-convention-namer';

import { WidgetStack } from './widget';

const app = new App();
const env = process.env.AWS_PROFILE || (process.env.CI ? 'usDev' : '');
const region = process.env.AWS_REGION || (process.env.CI ? 'us-east-1' : undefined);
if (!env) {
  console.log('You should probably set AWS_PROFILE before using this.');
}
const namedEnvFactory = core.Environment.findByName(env);
const namedEnv = namedEnvFactory(region);

console.log(\`Deploying to \${JSON.stringify(namedEnv.name)} in \${JSON.stringify(namedEnv.region)}.\`);

const commonProps = {
  businessUnit: core.BusinessUnit.PRODUCT,
  clickUpEnvironment: core.ClickUpEnvironment.PRODUCTION,
  clickUpRole: core.ClickUpRole.APP,
  confidentiality: core.Confidentiality.PUBLIC,
  namedEnv,
};
new WidgetStack(app, new Namer(['my', 'cool', 'widget']), {
  ...commonProps,
  managedPolicyName: 'yarf',
});

app.synth();
"
`;

exports[`ClickUpCdkTypeScriptApp defaults src/widget.ts 1`] = `
"
import { core } from '@time-loop/cdk-library';
import { aws_iam, aws_kms, RemovalPolicy } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { Namer } from 'multi-convention-namer';

export interface WidgetProps {
  /**
   * The name to assign the widget policy.
   * @default - have CDK generate a unique name
   */
  readonly managedPolicyName?: string;
}

/**
 * Useful comment describing the Widget construct.
 */
export class Widget extends Construct {
  // Expose the policy for use by another Construct in this Stack.
  // WARNING: cdk absolutely will let you pass objects between stacks.
  // This will generate CfnOutputs that will in turn create tight coupling between stacks.
  // This is almost never a Good Idea.
  // Instead, consider using \`core.Param.put\` to put the ARN
  // and then use myThingy.fromThingyArn() to instantiate a Interface object.
  readonly policy: aws_iam.ManagedPolicy;

  constructor(scope: Construct, id: Namer, props: WidgetProps) {
    super(scope, id.pascal);

    const key = new aws_kms.Key(this, 'Key', {
      removalPolicy: RemovalPolicy.DESTROY,
    });
    // Save the key ARN using SSM, for use by another Stack
    core.Param.put(key, 'keyArn', { rootId: 'ThisAppName' });

    this.policy = new aws_iam.ManagedPolicy(this, 'Policy', {
      managedPolicyName: props.managedPolicyName,
    });

    // Pull the distributionId from an SSM parameter
    const distributionId = core.Param.get(this, 'distributionId', {
      rootId: 'SomeOtherApp',
      stackId: 'SomeOtherStack',
      constructId: 'FooBar',
    });

    // Note that the \`allow()\` is implicit for PolicyStatements
    this.policy.addStatements(
      // Reference the ARN of a locally created thing.
      new aws_iam.PolicyStatement({ sid: 'DescriptiveName', actions: ['kms:Decrypt*'], resources: [key.keyArn] }),
      // Reference the distributionId from an SSM parameter
      new aws_iam.PolicyStatement({
        sid: 'GrantInvalidation',
        actions: ['cloudfront:CreateInvalidation'],
        resources: [\`arn:aws:cloudfront::\${distributionId}:*\`],
      }),
    );
  }
}

export interface WidgetStackProps extends core.StackProps, WidgetProps {}

export class WidgetStack extends core.Stack {
  constructor(scope: Construct, id: Namer, props: WidgetStackProps) {
    super(scope, id, props);

    const widget = new Widget(this, id, props);

    // Example of passing the widget.policy to another construct in the same stack.
    // \`\`\`
    // new FrozzleBop(this, id, { policy: widget.policy });
    // \`\`\`
    // placeholder to keep lint happy
    console.log(widget.policy.managedPolicyArn);
  }
}
"
`;

exports[`ClickUpCdkTypeScriptApp defaults test/widget.test.ts 1`] = `
"import { core } from '@time-loop/cdk-library';
import { App, assertions } from 'aws-cdk-lib';
import { Namer } from 'multi-convention-namer';
import { WidgetStack } from '../src/widget';

// Minimum props required by @time-loop/cdk-library/core.StackProps
const commonProps = {
  businessUnit: core.BusinessUnit.PRODUCT,
  clickUpEnvironment: core.ClickUpEnvironment.PRODUCTION,
  clickUpRole: core.ClickUpRole.APP,
  confidentiality: core.Confidentiality.PUBLIC,
  namedEnv: core.Environment.usDev('us-west-2'),
};

describe('Widget', () => {
  describe('default', () => {
    // These are resources shared by all the tests in this context.
    const app = new App();
    const stack = new WidgetStack(app, new Namer(['test']), commonProps);
    const template = assertions.Template.fromStack(stack);

    // Tests are super thin, consisting of just an assertion.
    test('creates resources', () => {
      ['AWS::IAM::ManagedPolicy', 'AWS::KMS::Key'].forEach((resource) => template.resourceCountIs(resource, 1));
    });

    // Demonstrate super-cool matcher stuff
    test('policy should reference key', () => {
      template.hasResourceProperties('AWS::IAM::ManagedPolicy', {
        PolicyDocument: {
          // The statement array must contain the following array.
          Statement: assertions.Match.arrayWith([
            assertions.Match.objectLike({
              // The array must contain an object with at least the following key / value.
              Resource: {
                'Fn::GetAtt': [
                  assertions.Match.anyValue(), // TODO: figure out how to actually reference the generated Kms key
                  'Arn',
                ],
              },
            }),
          ]),
        },
      });
    });
  });

  describe('options', () => {
    // Here we aren't sharing setup code...
    test('managedPolicyName', () => {
      // ...because each test is exercising a specific part of the functionality.
      // Which means that setups are slightly different and we can't re-use things.
      const app = new App();
      const stack = new WidgetStack(app, new Namer(['test']), {
        ...commonProps,
        managedPolicyName: 'fakeName',
      });
      const template = assertions.Template.fromStack(stack);

      template.hasResourceProperties('AWS::IAM::ManagedPolicy', {
        ManagedPolicyName: 'fakeName',
      });
    });
  });
});
"
`;
